src/Bounds.cpp

```diff
@@ -1039,11 +1039,29 @@ class Bounds : public IRVisitor {
                     if (op->is_intrinsic(Call::shift_left)) {
                         if (t.is_int() && t.bits() >= 32) {
                             // Overflow is UB
-                            if (a_interval.has_lower_bound() && b_interval.has_lower_bound() && can_prove(b_interval.min >= 0 && b_interval.min < t.bits())) {
+                            if (a_interval.has_lower_bound() &&
+                                b_interval.has_lower_bound() &&
+                                can_prove(b_interval.min >= 0 &&
+                                          b_interval.min < t.bits())) {
                                 interval.min = a_interval.min << b_interval.min;
+                            } else if (a_interval.has_lower_bound() &&
+                                       b_interval.has_lower_bound() &&
+                                       can_prove(b_interval.min < 0 &&
+                                                 !b_interval.min.type().is_uint() &&
+                                                 cast<int>(b_interval.min) > -t.bits())) {
+                                interval.min = a_interval.min >> abs(b_interval.min);
                             }
-                            if (a_interval.has_upper_bound() && b_interval.has_upper_bound() && can_prove(b_interval.max >= 0 && b_interval.max < t.bits())) {
+                            if (a_interval.has_upper_bound() &&
+                                b_interval.has_upper_bound() &&
+                                can_prove(b_interval.max >= 0 &&
+                                          b_interval.max < t.bits())) {
                                 interval.max = a_interval.max << b_interval.max;
+                            } else if (a_interval.has_upper_bound() &&
+                                       b_interval.has_upper_bound() &&
+                                       can_prove(b_interval.max < 0 &&
+                                                 !b_interval.max.type().is_uint() &&
```

Comment:
This is a constant so I think it should go outside (before) the can_prove (there's no need to make the simplifer deal with this, it's probably not a problem but it also makes the code clearer IMO).

Reply:
Done